### ISSUE TITLE
locale_gen fails with python3

### ISSUE THREAD

**Issue by gnotari** [ansible/ansible#19426]
Ansible 2.2.0.0 on osx Sierra (brew install)
pipelining = True

passed option -e 'ansible_python_interpreter=/usr/bin/python3'

playbook runs ok with python2 (installed in a pre-task, as the target is Ubuntu 16.04)

`fatal: [192.168.239.194]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Shared connection to 192.168.239.194 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n File \"/tmp/ansible_lhviwmqc/ansible_module_locale_gen.py\", line 239, in <module>\r\n main()\r\n File \"/tmp/ansible_lhviwmqc/ansible_module_locale_gen.py\", line 217, in main\r\n if is_present(name):\r\n File \"/tmp/ansible_lhviwmqc/ansible_module_locale_gen.py\", line 96, in is_present\r\n return any(fix_case(name) == fix_case(line) for line in output.splitlines())\r\n File \"/tmp/ansible_lhviwmqc/ansible_module_locale_gen.py\", line 96, in <genexpr>\r\n return any(fix_case(name) == fix_case(line) for line in output.splitlines())\r\n File \"/tmp/ansible_lhviwmqc/ansible_module_locale_gen.py\", line 101, in fix_case\r\n for s, r in LOCALE_NORMALIZATION.iteritems():\r\nAttributeError: 'dict' object has no attribute 'iteritems'\r\n", "msg": "MODULE FAILURE"}`

**Comment by robinro** [comment]
I believe this is fixed in the devel branch. Can you please try with the devel version?

**Comment by gnotari** [comment]
sadly no, I get a different error
This time I'm using Ansible 2.3 (devel) from a fresly created docker (alpine linux)

`fatal: [192.168.239.194]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Shared connection to 192.168.239.194 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_hjvbi4qf/ansible_module_locale_gen.py\", line 246, in <module>\r\n    main()\r\n  File \"/tmp/ansible_hjvbi4qf/ansible_module_locale_gen.py\", line 223, in main\r\n    if is_present(name):\r\n  File \"/tmp/ansible_hjvbi4qf/ansible_module_locale_gen.py\", line 102, in is_present\r\n    return any(fix_case(name) == fix_case(line) for line in output.splitlines())\r\n  File \"/tmp/ansible_hjvbi4qf/ansible_module_locale_gen.py\", line 102, in <genexpr>\r\n    return any(fix_case(name) == fix_case(line) for line in output.splitlines())\r\n  File \"/tmp/ansible_hjvbi4qf/ansible_module_locale_gen.py\", line 108, in fix_case\r\n    name = name.replace(s, r)\r\nTypeError: a bytes-like object is required, not 'str'\r\n", "msg": "MODULE FAILURE"}`

**Comment by westonplatter** [comment]
This issue was addressed by https://github.com/ansible/ansible-modules-extras/pull/3459, but I'm still having issues too.

__example setup__,

ansible version,
```sh
ansible --version
ansible 2.3.0
  config file = /path/to/my_ansible_playbooks_repo/ansible.cfg
  configured module search path = Default w/o overrides
```

hosts file,
```yml
[myhosts:children]
production

[production]
1.1.1.1

[production:vars]
ansible_python_interpreter=/usr/bin/python3
env=production
rails_env=production
```

playbook,
```yml
---
- name: server configure
  hosts: myhosts
  user: ubuntu
  gather_facts: false
  tasks:
    - name: locale en_US.UTF-8
      become_user: root
      become: yes
      locale_gen: name=en_US.UTF-8 state=present
```

playbook kicked off via,
```sh
ansible-playbook    -i hosts/hostfile   --vault-password-file ~/.ansiblevault_password.txt  server.yml --verbose
```

resulting output,
```sh
TASK [locale en_US.UTF-8] *****************************************************************
fatal: [1.1.1.1]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Connection to 1.1.1.1 closed.\r\n", "module_stdout": "\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_67z5z0ua/ansible_module_locale_gen.py\", line 246, in <module>\r\n    main()\r\n  File \"/tmp/ansible_67z5z0ua/ansible_module_locale_gen.py\", line 223, in main\r\n    if is_present(name):\r\n  File \"/tmp/ansible_67z5z0ua/ansible_module_locale_gen.py\", line 102, in is_present\r\n    return any(fix_case(name) == fix_case(line) for line in output.splitlines())\r\n  File \"/tmp/ansible_67z5z0ua/ansible_module_locale_gen.py\", line 102, in <genexpr>\r\n    return any(fix_case(name) == fix_case(line) for line in output.splitlines())\r\n  File \"/tmp/ansible_67z5z0ua/ansible_module_locale_gen.py\", line 108, in fix_case\r\n    name = name.replace(s, r)\r\nTypeError: a bytes-like object is required, not 'str'\r\n", "msg": "MODULE FAILURE", "rc": 0}
	to retry, use: --limit @/my/path/server_playbook.retry
```

**Comment by abadger** [comment]
I've just submitted #19854 which I think will fix the latest error you are reporting.  If one of you could test and see if there are more errors after that it would be appreciated.

**Comment by robinro** [comment]
This shouts "I want an integration test" to me...

As a start maybe
```
- locale_gen: name=en_US.UTF-8 state=present
- locale_gen: name=pt_BR state=present
```

**Comment by abadger** [comment]
<nod>  Unfortunately, I believe there's also a platform requirement.  I can't remember for sure but perhaps it only works on Ubuntu and Debian machines (with separate code paths for each :-( ?

**Comment by robinro** [comment]
Many (most?) modules have platform requirements: package managers, ... That doesn't stop tests.

**Comment by abadger** [comment]
<nod>  I was tired.. I meant, "The integration test will need more than just those lines" rather than "we can't write an integration test for this".  I've added some things to the PR... but it doesn't look like the tests are being run yet.  I'll ask mattclay if I'm doing something wrong.

**Comment by abadger** [comment]
Fixed in devel and stable-2.2 branch.  And btw, the issue was that I was dumb.  Forgot to git add the new locale_gen test directory.

**Comment by westonplatter** [comment]
thanks @abadger the help. I really appreciate you looking into the issue!
